experimental/QuantumAnalogs/src/QuantumAnalogs.jl

```diff
@@ -0,0 +1,206 @@
+################################################################################
+# Quantum analogs
+#
+# Copyright (C) 2020 Ulrich Thiel, ulthiel.com/math
+#
+# Mar 2023: Imported into OSCAR by UT
+#
+# Future features could include:
+# - Symmetric quantum numbers
+# - Two-colored quantum numbers
+################################################################################
+
+export quantum_integer, quantum_factorial, quantum_binomial
+
+################################################################################
+# Quantum integers
+################################################################################
+@doc raw"""
+    quantum_integer(n::IntegerUnion, q::RingElement)
+    quantum_integer(n::IntegerUnion)
+
+Return the quantum integer $[n]_q$ which is defined as $\frac{q^n-1}{q-1}$
+when ``q-1`` is invertible.
+
+For general ring elements `q`, we use the following identities to compute
+$[n]_q$: if `n` is non-negative, then $[n]_q = \sum_{i=0}^{n-1} q^i$. To
+handle negative values `n` we use the identity $[n]_q = -q^{n} [-n]_q$. Thus for
+negative `n` we require `q` to be invertible.
+
+Note that for ``q=1`` we obtain $[n]_1 = n$ hence the quantum integers are
+"deformations" of the usual integers. For details about these objects see
+[Con00](@cite) or [KC02](@cite).
+
+If `q` is omitted then it defaults to the generator of a Laurent polynomial
+ring over the integers.
+
+# Examples
+```jldoctest
+julia> quantum_integer(3)
+q^2 + q + 1
+
+julia> quantum_integer(-3)
+-q^-1 - q^-2 - q^-3
+
+julia> quantum_integer(3,2)
+7
+
+julia> quantum_integer(-3,2)
+ERROR: DomainError with -3:
+Cannot raise an integer x to a negative power -3.
+
+julia> quantum_integer(-3,2//1)
+-7//8
+
+julia> K,i = cyclotomic_field(4, "i");
+
+julia> quantum_integer(3, i)
+i
+```
+"""
+function quantum_integer(n::IntegerUnion, q::RingElement)
+  R = parent(q)
+  isone(q) && return R(n)
+  n < 0 && return -q^n * quantum_integer(-n, q)
+
+  z = zero(R)
+  qi = one(R)
+  for i = 0:n-1
+    # at this point qi = q^i
+    z = add!(z, qi)
+    qi = mul!(qi, q)
+  end
+  return z
+end
+
+function quantum_integer(n::IntegerUnion)
+  R,q = laurent_polynomial_ring(ZZ, "q")
```

Comment:
```suggestion
  R,q = laurent_polynomial_ring(ZZ, "q"; cached=false)
```
This shouldn't cache the poly ring by our style guide. However, for interactive use, it is incredibly unuseful if this creates new Laurent rings every time you call this function, as the results are not compatible

Reply:
Related to this is the introduction of a parent (quantum integer ring). This would in my opinion be the cleaner solution, but means more work.

Reply:
I think the idea of a "parent quantum integer ring" sounds intriguing at first but in practice it wasn't so clear how it should work and who would be using it. I'd rather net let this PR be delayed over this.

Reply:
And yeah I deliberately did not add the `cached=false` because it seems extremely unhelpful for its intended use.